src/Compiler/Service/ServiceParseTreeWalk.fs

```diff
@@ -1103,9 +1114,478 @@ module SyntaxTraversal =
+
+                            let rec loop =
+                                function
+                                | [] -> defaultTraverse synModuleSigDecl
+                                | ty :: types ->
+                                    match folder state path (SyntaxNode.SynTypeDefnSig ty) with
+                                    | Some state' ->
+                                        state <- state'
+                                        loop types
+                                    | None -> Some()
+
+                            loop types
+
+                        | _ -> defaultTraverse synModuleSigDecl
+
+                    | None -> Some()
+
+                member _.VisitValSig(path, defaultTraverse, valSig) =
+                    match path with
+                    | SyntaxNode.SynMemberSig _ as parent :: path ->
+                        match folder state path parent with
+                        | Some state' ->
+                            match folder state' path (SyntaxNode.SynValSig valSig) with
+                            | Some state' ->
+                                state <- state'
+                                defaultTraverse valSig
+                            | None -> Some()
+                        | None -> Some()
+                    | _ ->
+                        match folder state path (SyntaxNode.SynValSig valSig) with
+                        | Some state' ->
+                            state <- state'
+                            defaultTraverse valSig
+                        | None -> Some()
+
+                member _.VisitSimplePats(path, _pat) =
+                    match path with
+                    | SyntaxNode.SynMemberDefn _ as node :: path ->
+                        match folder state path node with
+                        | Some state' ->
+                            state <- state'
+                            None
+                        | None -> Some()
+                    | _ -> None
+
+                member _.VisitInterfaceSynMemberDefnType(path, _synType) =
+                    match path with
+                    | SyntaxNode.SynMemberDefn _ as node :: path ->
+                        match folder state path node with
+                        | Some state' ->
+                            state <- state'
+                            None
+                        | None -> Some()
+                    | _ -> None
+            }
+
+        ignore<unit option> (SyntaxTraversal.traverseUntil pick pos visitor ast)
+        state
+
+    let foldWhile folder state (ast: Ast) =
+        let pickAll _ _ _ diveResults =
+            let rec loop =
+                function
+                | [] -> None
+                | (_, project) :: rest ->
+                    ignore (project ())
+                    loop rest
+
+            loop diveResults
+
+        let m = (range0, ast) ||> List.fold (fun acc node -> unionRanges acc node.Range)
```

Comment:
Perhaps also outside of the scope of this PR, but it is a shame that `ParsedInput.Range` returns such a poor result.

https://github.com/dotnet/fsharp/blob/e2496896c128ccfde33c92f45bbe0d2aa738873a/src/Compiler/SyntaxTree/SyntaxTree.fs#L1781-L1785

This should really reflect the file from the first token till the last. Trivia included.

Reply:
In Fantomas, we work around this using [some custom code](https://github.com/fsprojects/fantomas/blob/d8f44dbe6c47bfb6371aed1dce9ee5227bf773c8/src/Fantomas.Core/ASTTransformer.fs#L3459-L3549).

Reply:
> In Fantomas, we work around this using [some custom code](https://github.com/fsprojects/fantomas/blob/d8f44dbe6c47bfb6371aed1dce9ee5227bf773c8/src/Fantomas.Core/ASTTransformer.fs#L3459-L3549).

As an aside, I see some examples in that file where you refer to `ParsedInput` as `ast`, which helps me understand why you might want to avoid introducing the new `Ast = SyntaxNode list` alias :)

Reply:
I actually think I want to die on this hill: AST stands for Abstract Syntax **Tree** and should never ever be the abbreviation for anything else than a tree.

Reply:
I do agree that `ParsedInput.Range` being only the range of the first module/namespace in the file is odd. I _think_ that unioning the ranges of the top-level modules/namespaces in a file is good enough for what this is doing here.